### PR TITLE
Do not register device on each app resume

### LINKED ISSUE #4932: Do not register FCM token on app resume
Currently, we run `FCMRegistrationIntentService#enqueueWork` on every single `AppInitializer#onAppComesFromBackground` which happens on `Activity#onResume`. 

This is probably unwanted behaviour as we flood our backend and influence clients throughput. Also, we started to receive `TOO_MANY_REGISTRATION` error from Firebase.

### PR DESCRIPTION
**Pull request by wzieba**
Closes: #4932

Corresponding FluxC PR: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2399 (has to be merged first).

### Description

This PR changes the logic of how we approach registering devices on WPCOM in order to start receiving notifications.

In current `trunk` version, we call `FCMRegistrationIntentService#enqueueWork` on every `AppInitializer#onAppComesFromBackground` invocation, which makes additional requests to FCM and WPCOM APIs.

So now, there are three scenarios when we want to send an FCM token:
A) User logged in
B) The device token has been regenerated (e.g. due to security reasons)
C) User opened the app and their device is not registered (e.g. during the log in backend was unresponsive/internet connection was not stable etc.)

and we should carefully test each case in this PR.

### Testing instructions

Some tips:
1. I'd recommend using the `quicklogin` script - it'll be much more pleasant to execute those test cases.
2. Have a second device (iOS/Android) with correctly working notifications next to you to compare. Sometimes notifications are not sent right away, it's good to have confirmation if there's a problem with this PR or just if there's some delay.
3. Please repeat the test cases if possible. Notifications are a crucial feature for Woo mobile users.

##### TC 1 Log in
1. Open the app
1. Log in and select a site
1. **Assert that the device receives notification**

##### TC 2 Log in as a different user
- [x] Done 
1. Perform TC 1
2. Log out
3. Log in as a different Jetpack user (having a different set of stores)
4. **Assert that the device receives notifications from the second store and does not receive notifications from the first store**

##### TC 3 Check if multiple stores notifications still work
- [x] Done 
1. Log in with an account that has more than 2 stores
3. **Assert that the device receives notifications from the not-logged in-store**

##### TC 4 Assert that we stopped registering devices on each app resume
- [x] Done 
1. Log in to the app
2. Open Flipper (or any other tool for tracking app network traffic)
3. Go to the background and open the app several times
4. **Assert that you can't see multiple calls to `public-api.wordpress.com/rest/v1/devices/new/`**

##### TC 5 Regression check
- [x] Done
1. Install version from `trunk`
2. Log in
3. **Assert that notifications are received**
4. Install version from this branch
5. **Assert that notifications are still received**

##### TC6 App sends token if its regenerated
- [x] Done
1. Execute 
```
adb shell am broadcast -a com.google.firebase.messaging.NEW_TOKEN -n com.woocommerce.android.dev/com.google.firebase.iid.FirebaseInstanceIdReceiver --es "token" "newtoken"
```
2. **Assert  that you can see a call to `public-api.wordpress.com/rest/v1/devices/new/` with a payload containing "newtoken"**

##### TC 7 App sends token if the app is not registered
- [x] Done
1. Perform TC 1
2. Go to Flipper -> SharedPreferencesViewer
7. Select `(...)_fluxc-preferences`
8. Remove `NOTIFICATIONS_SERVER_ID_PREF_KEY`
9. Close the app
10. Run the app
4. **Assert that you can see a call to `public-api.wordpress.com/rest/v1/devices/new/` and that the device receives notifications**

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt [hunk 1]
```diff
@@ -147,6 +150,14 @@ class AppInitializer @Inject constructor() : ApplicationLifecycleListener {
...
             siteObserver.observeAndUpdateSelectedSiteData()
         }
+
+        appCoroutineScope.launch {
+            getDeviceRegistrationStatus().apply {
+                if (this == GetDeviceRegistrationStatus.Status.UNREGISTERED) {
+                    registerDevice()
+                }
+            }
+        }
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt]: Scenario C)

File: WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt [hunk 2]
```diff
@@ -288,6 +294,12 @@ class AppInitializer @Inject constructor() : ApplicationLifecycleListener {
                 AnalyticsTracker.sendUsageStats = !accountStore.account.tracksOptOut
             }
         }
+
+        if (!isLoggedOut) {
+            appCoroutineScope.launch {
+                registerDevice()
+            }
+        }
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt]: Scenario A)

File: WooCommerce/src/main/kotlin/com/woocommerce/android/push/FCMMessageService.kt [hunk 3]
```diff
@@ -2,24 +2,26 @@ package com.woocommerce.android.push
...
-     * the previous token had been compromised. Note that this is called when the
-     * FCM registration token is initially generated so this is where you would retrieve the token.
-     */
     override fun onNewToken(newToken: String) {
-        WooLog.d(T.NOTIFS, "Sending FCM token to our remote services: $newToken")
-        notificationRegistrationHandler.onNewFCMTokenReceived(newToken)
+        appPrefsWrapper.setFCMToken(newToken)
+        CoroutineScope(Dispatchers.IO).launch {
+            registerDevice()
+        }
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/push/FCMMessageService.kt]: Scenario B)

File: WooCommerce/src/main/kotlin/com/woocommerce/android/push/FCMMessageService.kt [hunk 4]
```diff
@@ -2,24 +2,26 @@ package com.woocommerce.android.push
...
 
-    /**
-     * Called if the FCM registration token is updated. This may occur if the security of
-     * the previous token had been compromised. Note that this is called when the
-     * FCM registration token is initially generated so this is where you would retrieve the token.
-     */
     override fun onNewToken(newToken: String) {
-        WooLog.d(T.NOTIFS, "Sending FCM token to our remote services: $newToken")
-        notificationRegistrationHandler.onNewFCMTokenReceived(newToken)
+        appPrefsWrapper.setFCMToken(newToken)
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/push/FCMMessageService.kt:28]: We save FCM token no matter what because it's always _valid_.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/AppUpdateReceiver.kt [hunk 5]
```diff
@@ -1,20 +0,0 @@
...
-import android.content.BroadcastReceiver
-import android.content.Context
-import android.content.Intent
-import com.woocommerce.android.push.FCMRegistrationIntentService
-import com.woocommerce.android.util.WooLog
-import com.woocommerce.android.util.WooLog.T
-
-/**
- * This is needed temporary to force re-registering the token with backend, to update the `selected_blog_id`
- */
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/AppUpdateReceiver.kt:12]: If I'm correct, it's not needed anymore right? We either way previously were registering devices on each app resume so everyone who had, already migrated to null `selected_blog_id` (needed for multiple stores notifications).

File: WooCommerce/src/main/kotlin/com/woocommerce/android/AppPrefs.kt [hunk 6]
```diff
@@ -204,14 +205,10 @@ object AppPrefs {
...
 
-    fun getFCMToken() = getString(DeletableSitePrefKey.WC_PREF_NOTIFICATIONS_TOKEN)
+    fun getFCMToken() = getString(UndeletablePrefKey.WC_PREF_NOTIFICATIONS_TOKEN)
 
     fun setFCMToken(token: String) {
-        setString(DeletableSitePrefKey.WC_PREF_NOTIFICATIONS_TOKEN, token)
-    }
-
-    fun removeFCMToken() {
-        remove(DeletableSitePrefKey.WC_PREF_NOTIFICATIONS_TOKEN)
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/AppPrefs.kt:214]: There's no need to remove the FCM token when the user logs out. The FCM token is assigned to the device, not Woo/WPCOM user account. 

So if we'll remove the token and then ask FCM for a token, we'll receive the same token.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/push/FCMMessageService.kt (hunk 3 above)
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/push/FCMMessageService.kt]: Is this the correct scope? Should I define a `Job` and cancel it in `onDestroy` of the Service? I'm not sure, because I don't want to cancel this coroutine ever. But what will happen if `onDestroy` happens (and when can it happen? 🤔 ). I'm not sure about all of those.

**Comment by wzieba** [comment]
Hi @atorresveiga @anitaa1990 👋 

@atorresveiga could you please review this PR and perform test cases? I especially look if I didn't forget about any corner cases that would break receiving notifications.

@anitaa1990 - do you have the capacity to also browse the code? You've been working on multiple store notifications so maybe you can spot if I forgot about some case where this solution won't work?

Thank you 🙏 .

File: WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt (hunk 2 above)
Comment (anitaa1990) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt]: ❓ I'm curious to understand why this logic is not added to the [onFirstActivityResumed](https://github.com/woocommerce/woocommerce-android/blob/issue/4932_do_not_request_token_on_app_resume/WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt#L179)? We already check in that method if the user is logged in and we fetch the account details in this method. Is there a need to wait till the account details are updated before registering the device? (By the way, there could be, I'm not sure 😅 My question is more out of curiosity)

Comment (anitaa1990) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt]: ❓ It looks like the [RegisterDevice](https://github.com/woocommerce/woocommerce-android/blob/b03470338b61bd7cf7ee99217051454c7f110506/WooCommerce/src/main/kotlin/com/woocommerce/android/push/RegisterDevice.kt#L14) function already checks in the user is logged in so do we need another check here?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt (hunk 1 above)
Comment (anitaa1990) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt]: ❓ We could maybe just use this for Scenario C and A? WDYT?

**Review by anitaa1990 — COMMENTED** [review]
Really nice work on this @wzieba! I didn't really test the changes in the PR yet but reviewed the code and everything looks good 👍  I left a couple of questions in the code but that is just to understand your opinions here 😄

**Comment by wzieba** [comment]
Thank you @anitaa1990 for the review! ~Just FYI I'm working on improvements now, I'll try the approach with `onFirstActivityResumed`~. Answered in https://github.com/woocommerce/woocommerce-android/pull/6535#discussion_r879442071

File: WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt (hunk 2 above)
Comment (wzieba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt]: > I'm curious to understand why this logic is not added to the [onFirstActivityResumed](https://github.com/woocommerce/woocommerce-android/blob/issue/4932_do_not_request_token_on_app_resume/WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt#L179)?

I tested this, but unfortunately, we can't move the check there as `onFirstActivityResumed` is not called when the app is resumed so e.g. in the case of TC 7 scenario, we won't send key again (if sending key previously failed).

> It looks like the [RegisterDevice](https://github.com/woocommerce/woocommerce-android/blob/b03470338b61bd7cf7ee99217051454c7f110506/WooCommerce/src/main/kotlin/com/woocommerce/android/push/RegisterDevice.kt#L14) function already checks in the user is logged in so do we need another check here?

Good point but `RegisterDevice` checks for access token and here we have check for `isLoggedOut` - I believe those two checks are not the same as we check them both in line 279.

I think it's safe to leave this additional check - e.g. there might be an error in logging out and this might propagate `OnAccountChanged` with errors (not sure but I believe it's possible). Without the `is logged out` check, we'll unnecessarily re-send the device token.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt (hunk 1 above)
Comment (wzieba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt]: (oh sorry I forgot to answer here)

We can't do this as `AppInitializer#init` will only run once for the application instance. So in the case of scenario A we won't send the key.

**Comment by atorresveiga** [comment]
Thanks for working on this, @wzieba! With these changes, the app will be more competent in calling the API and use the device resources wisely. Great work!

When testing TC4 and TC7, I noticed that we were calling the API Twice, the first time to fetch the merchant account and the second time to fetch the store settings. Here is a video:

https://user-images.githubusercontent.com/18119390/170390267-f35c96d2-aa29-439d-844d-4dfc29420ba1.mp4

A possible solution could be to move the `getDeviceRegistrationStatus()` check into the `registerDevice()` and only call the API when the device status is UNREGISTERED. The only problem here is that if the device token has been regenerated, we will need to mark it as  UNREGISTERED (delete the deviceId in the preferences) before calling `registerDevice()` otherwise will not send the request. I'm not saying this is a great solution. I just wanted to share this idea with you to see what you think.

I also found very interesting ideas [here](https://firebase.google.com/docs/cloud-messaging/manage-tokens) that maybe we can adopt in the future.

> your app should retrieve this token at the initial startup and save it to your app server alongside a timestamp.

> make sure to update tokens periodically.

I'm having some issues with **TC6** because I do not see the expected result. I will try again tomorrow and let you know.

Excellent work with the documentation of the test cases! Huge thanks  🙏🏻

File: WooCommerce/src/main/kotlin/com/woocommerce/android/push/FCMMessageService.kt (hunk 3 above)
Comment (atorresveiga) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/push/FCMMessageService.kt]: >But what will happen if onDestroy happens

I think that if `onDestroy` is called while the `registerDevice()` is running, the Service will be leaked for the time it takes for `registerDevice()` to complete. As pointed out by you and recommended [here](https://stackoverflow.com/questions/67847118/should-i-use-coroutine-in-onmessagereceived-of-firebasemessagingservice), we could create a job and cancel in onDestroy
```
val job = SupervisorJob()

 CoroutineScope(job).launch { 
   // Your Stuff here 
}

override fun onDestroy() {
    job.cancel()
    super.onDestroy()
}
```

or use the appCoroutineScope `@Inject @AppCoroutineScope lateinit var appCoroutineScope: CoroutineScope`

Comment (atorresveiga) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/push/FCMMessageService.kt]: What happens if something goes wrong while calling `registerDevice()` here? 🤔
Imagine this scenario:

1. The app has a valid token (the status is REGISTERED)
2. Firebase regenerates the token (e.g., due to security reasons)
3. Something goes wrong while updating the token

How can we recover from this scenario?
I believe the solution I share below of removing the deviceId from the preferences (marking the token as UNREGISTERED) could work. However, a better idea could be to use WorkManager and enqueue a work to update the token (I'm thinking out loud here) WDYT?

Comment (wzieba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/push/FCMMessageService.kt]: > we could create a job and cancel in onDestroy

Thanks - I did that.

> What happens if something goes wrong while calling registerDevice() here? 

Good point! To resolve this issue, I've introduced [`InvalidateDeviceRegistration`](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2399/files#diff-19007ce406c42cdd45a58cc8efc963dc02c7d1afcb91ec598e29d4b6d9cbe53bR6) which removes server id from local storage. This way, we make sure that Woo will see `UNREGISTERED` state and will try to approach sending the token again.

The only cons of this approach is that we don't unregister an old FCM token from our backend. We didn't do that before though and this PR is already getting big - I'd suggest leaving this for future improvements. Also - I have to admit that I _feel_ like this token regeneration happens rarely - I don't have any data to prove that though 😓 .

**Comment by wzieba** [comment]
Thank you for the review @atorresveiga ! I have to admit that I misinterpreted `onAccountChanged` - I thought it's propagated only when user account is changed, but it appears that it's also propagated on the app start. 

I've added improvements in https://github.com/woocommerce/woocommerce-android/pull/6535/commits/f2fb6ae51eaea7ebbb3ee3dc8bb777c77ccfa7db .

---

For TC 6 - I didn't mention this in the description (sorry!), but I think it'll work only on an emulator and on this one without Google APIs.

**Review by atorresveiga — APPROVED** [review]
Great work Wojtek! I've tested the new changes and everything is working great. :shipit:  🙌

File: WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt [hunk 7]
```diff
@@ -162,13 +164,12 @@ class AppInitializer @Inject constructor() : ApplicationLifecycleListener {
             application.registerReceiver(connectionReceiver, IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION))
         }
 
-        if (isGooglePlayServicesAvailable(application)) {
-            // Register for Cloud messaging
-            FCMRegistrationIntentService.enqueueWork(application)
-        }
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt:168]: @wzieba sorry for reviving this old PR 😅.

I was adding some stuff to the AppInitializer class, and I noticed that `isGooglePlayServicesAvailable` is not used anymore, before going forward with deleting it I wanted to check why it's not used now and it was used before, and this led me to this PR.

I didn't review the whole code, so I don't know all the changes, and that's why I'm asking you, was this removal of GMS check intentional?

Comment (wzieba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt:168]: 👋 thanks for checking 😃! TLDR; the removal of Google Play Services was intentional.

Before this PR, we were requesting a new FCM token at the app start. We had a check not to request FCM token if there were no Google Play Services, as we were sure it'd fail.

After this PR, we observe the result for FCM API - see `FCMMessageService#onNewToken`. As we do not request the token, we don't have to check for Google Play Services availability.

Does it put some more light into this part of the code?

Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt:168]: Thanks @wzieba, this is clear.
So we can safely remove isGooglePlayServicesAvailable as it's not needed anymore, I'll remove it as part of my other PR.
